File: app/src/main/java/com/waray/spendhound/AllTransactionsActivity.java
package com.waray.spendhound;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.Set;

public class AllTransactionsActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private RecentTransactionAdapter adapter;
    private ArrayList<RecentTransaction> transactionList;
    private Spinner monthSpinner;
    private TextView currentMonthTextView;
    private TextView transactionCountTextView;
    private ProgressBar loadingProgressBar;
    private LinearLayout emptyStateLayout;
    private FirebaseAuth mAuth;
    private String currentNickname = "";
    private List<String> availableMonths;
    private String selectedMonth;

    // Status Tabs
    private TextView allTabTV, paidTabTV, unpaidTabTV, pendingTabTV;
    private String selectedStatusTab = "All";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_all_transactions);

        mAuth = DeclareDatabase.getAuth();
        transactionList = new ArrayList<>();
        availableMonths = new ArrayList<>();

        initViews();
        getCurrentNickname();
    }

    private void initViews() {
        recyclerView = findViewById(R.id.allTransactionsRecyclerView);
        monthSpinner = findViewById(R.id.monthSpinner);
        currentMonthTextView = findViewById(R.id.currentMonthTextView);
        transactionCountTextView = findViewById(R.id.transactionCountTextView);
        loadingProgressBar = findViewById(R.id.loadingProgressBar);
        emptyStateLayout = findViewById(R.id.emptyStateLayout);

        // Status Tabs
        allTabTV = findViewById(R.id.allTabTV);
        paidTabTV = findViewById(R.id.paidTabTV);
        unpaidTabTV = findViewById(R.id.unpaidTabTV);
        pendingTabTV = findViewById(R.id.pendingTabTV);

        setupStatusTabs();

        adapter = new RecentTransactionAdapter(transactionList, this::onTransactionTap);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);
    }

    private void setupStatusTabs() {
        setStatusTabSelected(allTabTV);

        allTabTV.setOnClickListener(v -> {
            selectedStatusTab = "All";
            setStatusTabSelected(allTabTV);
            refreshTransactions();
        });

        paidTabTV.setOnClickListener(v -> {
            selectedStatusTab = "Paid";
            setStatusTabSelected(paidTabTV);
            refreshTransactions();
        });

        unpaidTabTV.setOnClickListener(v -> {
            selectedStatusTab = "Unpaid";
            setStatusTabSelected(unpaidTabTV);
            refreshTransactions();
        });

        pendingTabTV.setOnClickListener(v -> {
            selectedStatusTab = "Pending";
            setStatusTabSelected(pendingTabTV);
            refreshTransactions();
        });
    }

    private void setStatusTabSelected(TextView selectedTab) {
        // Reset all tabs
        allTabTV.setBackgroundResource(0);
        paidTabTV.setBackgroundResource(0);
        unpaidTabTV.setBackgroundResource(0);
        pendingTabTV.setBackgroundResource(0);

        // Set selected tab background
        selectedTab.setBackgroundResource(R.drawable.bg_status_tab_selected);
    }

    private void refreshTransactions() {
        if (selectedMonth != null) {
            fetchTransactionsForMonth(selectedMonth);
        }
    }

    private void onTransactionTap(RecentTransaction transaction) {
        // Handle tap if needed
    }

    private void getCurrentNickname() {
        String userId = Objects.requireNonNull(mAuth.getCurrentUser()).getUid();
        DatabaseReference userRef = DeclareDatabase.getDatabaseReference().child(userId);

        userRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    currentNickname = snapshot.child("username").getValue(String.class);
                    if (currentNickname == null) {
                        currentNickname = "";
                    }
                }
                loadAvailableMonths();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.e("AllTransactions", "Error getting nickname: " + error.getMessage());
                loadAvailableMonths();
            }
        });
    }

    private void loadAvailableMonths() {
        loadingProgressBar.setVisibility(View.VISIBLE);

        DatabaseReference transRef = DeclareDatabase.getDBRefTransaction();
        transRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Set<String> uniqueMonths = new HashSet<>();

                for (DataSnapshot monthSnapshot : dataSnapshot.getChildren()) {
                    String monthYear = monthSnapshot.getKey();
                    if (monthYear != null && !monthYear.isEmpty()) {
                        for (DataSnapshot daySnapshot : monthSnapshot.getChildren()) {
                            for (DataSnapshot timeSnapshot : daySnapshot.getChildren()) {
                                Transaction transaction = timeSnapshot.getValue(Transaction.class);
                                if (transaction != null && isUserInvolved(transaction, currentNickname)) {
                                    uniqueMonths.add(monthYear);
                                    break;
                                }
                            }
                            if (uniqueMonths.contains(monthYear)) break;
                        }
                    }
                }

                availableMonths = new ArrayList<>(uniqueMonths);
                Collections.sort(availableMonths, (m1, m2) -> {
                    try {
                        SimpleDateFormat sdf = new SimpleDateFormat("MMMM-yyyy", Locale.getDefault());
                        return sdf.parse(m2).compareTo(sdf.parse(m1));
                    } catch (Exception e) {
                        return m2.compareTo(m1);
                    }
                });

                setupMonthSpinner();
                loadingProgressBar.setVisibility(View.GONE);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.e("AllTransactions", "Error loading months: " + error.getMessage());
                loadingProgressBar.setVisibility(View.GONE);
            }
        });
    }

    private void setupMonthSpinner() {
        if (availableMonths.isEmpty()) {
            Calendar calendar = Calendar.getInstance();
            SimpleDateFormat dateFormat = new SimpleDateFormat("MMMM-yyyy", Locale.getDefault());
            availableMonths.add(dateFormat.format(calendar.getTime()));
        }

        SpinnerItemMonths spinnerAdapter = new SpinnerItemMonths(this, availableMonths);
        monthSpinner.setAdapter(spinnerAdapter);

        Calendar calendar = Calendar.getInstance();
        SimpleDateFormat dateFormat = new SimpleDateFormat("MMMM-yyyy", Locale.getDefault());
        String currentMonth = dateFormat.format(calendar.getTime());

        int defaultPosition = 0;
        for (int i = 0; i < availableMonths.size(); i++) {
            if (availableMonths.get(i).equals(currentMonth)) {
                defaultPosition = i;
                break;
            }
        }

        monthSpinner.setSelection(defaultPosition);
        selectedMonth = availableMonths.get(defaultPosition);
        updateMonthDisplay(selectedMonth);
        fetchTransactionsForMonth(selectedMonth);

        monthSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                String month = availableMonths.get(position);
                if (!month.equals(selectedMonth)) {
                    selectedMonth = month;
                    updateMonthDisplay(selectedMonth);
                    fetchTransactionsForMonth(selectedMonth);
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    private void updateMonthDisplay(String monthYear) {
        String displayMonth = monthYear.replace("-", " ");
        currentMonthTextView.setText(displayMonth);
    }

    @SuppressLint("NotifyDataSetChanged")
    private void fetchTransactionsForMonth(String monthYear) {
        loadingProgressBar.setVisibility(View.VISIBLE);
        emptyStateLayout.setVisibility(View.GONE);
        transactionList.clear();
        adapter.notifyDataSetChanged();

        DatabaseReference monthRef = DeclareDatabase.getDBRefTransaction().child(monthYear);

        monthRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @SuppressLint("NotifyDataSetChanged")
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                for (DataSnapshot daySnapshot : dataSnapshot.getChildren()) {
                    String day = daySnapshot.getKey();

                    for (DataSnapshot timeSnapshot : daySnapshot.getChildren()) {
                        Transaction transaction = timeSnapshot.getValue(Transaction.class);
                        String timeKey = timeSnapshot.getKey();

                        if (transaction != null && isUserInvolved(transaction, currentNickname)) {
                            
                            // Apply Status Filter
                            if (!matchesStatusFilter(transaction, selectedStatusTab)) {
                                continue;
                            }

                            String[] parts = monthYear.split("-");
                            String month = parts[0];
                            String year = parts.length > 1 ? parts[1] : "";
                            String displayDate = month + " - " + day;
                            String fullDateWithYear = month + " " + day + ", " + year;
                            String sortDateTime = year + "-" + month + "-" + day + " " + timeKey;

                            String transactionType = transaction.getTransactionType();
                            String details = transaction.getMultilineStr();
                            double paymentAmount = transaction.getPaymentAmount();
                            String paymentAmountStr = CurrencyUtils.formatAmountWithCurrency(paymentAmount);
                            int iconResource = getIconForTransactionType(transactionType);

                            java.util.List<String> payorsList = transaction.getPayorsDisplayNames();
                            if (payorsList == null || payorsList.isEmpty()) {
                                payorsList = transaction.getPayorsList();
                            }
                            java.util.List<String> payorUids = transaction.getPayorsList();
                            java.util.List<Double> amountsPaidList = transaction.getAmountsPaidList();
                            double totalIndividualPayment = transaction.getTotalIndividualPayment();

                            String createdBy = transaction.getPosterDisplayName();
                            if (createdBy == null || createdBy.isEmpty()) {
                                createdBy = transaction.getUsernamePost();
                            }
                            String createdByUid = transaction.getUsernamePost();

                            RecentTransaction recentTrans = new RecentTransaction(
                                    displayDate, transactionType, details, paymentAmountStr,
                                    iconResource, sortDateTime, payorsList, payorUids,
                                    amountsPaidList, totalIndividualPayment, fullDateWithYear,
                                    createdBy, createdByUid, monthYear, day, timeKey
                            );
                            transactionList.add(recentTrans);
                        }
                    }
                }

                Collections.sort(transactionList, (t1, t2) -> {
                    String dateTime1 = t1.getSortDateTime();
                    String dateTime2 = t2.getSortDateTime();
                    if (dateTime1 != null && dateTime2 != null) {
                        return dateTime2.compareTo(dateTime1);
                    }
                    return 0;
                });

                adapter.notifyDataSetChanged();
                loadingProgressBar.setVisibility(View.GONE);

                int count = transactionList.size();
                transactionCountTextView.setText(count + (count == 1 ? " transaction" : " transactions"));

                if (transactionList.isEmpty()) {
                    emptyStateLayout.setVisibility(View.VISIBLE);
                    recyclerView.setVisibility(View.GONE);
                } else {
                    emptyStateLayout.setVisibility(View.GONE);
                    recyclerView.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Log.e("AllTransactions", "Error loading transactions: " + error.getMessage());
                loadingProgressBar.setVisibility(View.GONE);
            }
        });
    }

    private boolean matchesStatusFilter(Transaction transaction, String statusFilter) {
        if ("All".equalsIgnoreCase(statusFilter)) return true;

        List<Double> paidAmounts = transaction.getAmountsPaidList();
        double totalToPay = transaction.getTotalIndividualPayment();

        if (paidAmounts == null || paidAmounts.isEmpty()) {
            return "Unpaid".equalsIgnoreCase(statusFilter);
        }

        boolean allPaid = true;
        boolean allUnpaid = true;

        for (Double paid : paidAmounts) {
            if (paid < totalToPay) {
                allPaid = false;
            }
            if (paid > 0) {
                allUnpaid = false;
            }
        }

        String status;
        if (allPaid) {
            status = "Paid";
        } else if (allUnpaid) {
            status = "Unpaid";
        } else {
            status = "Pending";
        }

        return status.equalsIgnoreCase(statusFilter);
    }

    private int getIconForTransactionType(String transactionType) {
        if ("Electricity".equals(transactionType)) return R.drawable.lightning_bolt;
        if ("Water".equals(transactionType)) return R.drawable.faucet;
        if ("Rent".equals(transactionType)) return R.drawable.house;
        if ("Internet".equals(transactionType)) return R.drawable.internet;
        if ("Online Shopping".equals(transactionType)) return R.drawable.online_shopping;
        if ("Travel".equals(transactionType)) return R.drawable.travel;
        if ("Groceries".equals(transactionType)) return R.drawable.groceries;
        if ("Foods".equals(transactionType)) return R.drawable.hamburger;
        if ("House Necessity".equals(transactionType)) return R.drawable.necessities;
        if ("Transportation".equals(transactionType)) return R.drawable.vehicles;
        return R.drawable.others;
    }

    private boolean isUserInvolved(Transaction transaction, String usernameOrUid) {
        if (transaction == null || usernameOrUid == null || usernameOrUid.isEmpty()) return false;
        String currentUid = Objects.requireNonNull(mAuth.getCurrentUser()).getUid();
        if (transaction.isUserInvolvedByUid(currentUid)) return true;
        if (usernameOrUid.equals(transaction.getUsernamePost())) return true;
        if (usernameOrUid.equals(transaction.getPosterDisplayName())) return true;
        java.util.List<String> payorsList = transaction.getPayorsList();
        if (payorsList != null && payorsList.contains(usernameOrUid)) return true;
        java.util.List<String> payorsDisplayNames = transaction.getPayorsDisplayNames();
        if (payorsDisplayNames != null && payorsDisplayNames.contains(usernameOrUid)) return true;
        return false;
    }
}

File: app/src/main/java/com/waray/spendhound/AllTransactionsActivity.kt
package com.waray.spendhound

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.datepicker.MaterialDatePicker
import androidx.core.util.Pair
import io.github.jan.supabase.gotrue.Auth
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

class AllTransactionsActivity : AppCompatActivity() {

    private var recyclerView: RecyclerView? = null
    private var adapter: RecentTransactionAdapter? = null
    private var transactionList: ArrayList<RecentTransaction> = ArrayList()
    private var dateRangeSpinner: android.widget.Spinner? = null
    private var transactionCountTextView: TextView? = null
    private var loadingProgressBar: ProgressBar? = null
    private var emptyStateLayout: LinearLayout? = null
    private var mAuth: Auth? = null
    private var currentNickname: String? = ""
    private var currentUserNumericId: Long? = null
    
    private var startDate: Long = 0L
    private var endDate: Long = 0L

    // Status Tabs
    private var allTabTV: TextView? = null
    private var paidTabTV: TextView? = null
    private var unpaidTabTV: TextView? = null
    private var pendingTabTV: TextView? = null
    private var selectedStatusTab = "All"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_transactions)

        mAuth = DeclareDatabase.auth
        
        initViews()
        getCurrentNickname()
        setupDateRangeSpinner()
    }

    private fun initViews() {
        recyclerView = findViewById(R.id.allTransactionsRecyclerView)
        dateRangeSpinner = findViewById(R.id.dateRangeSpinner)
        transactionCountTextView = findViewById(R.id.transactionCountTextView)
        loadingProgressBar = findViewById(R.id.rvSkeletonTransactions)
        emptyStateLayout = findViewById(R.id.emptyStateLayout)

        allTabTV = findViewById(R.id.allTabTV)
        paidTabTV = findViewById(R.id.paidTabTV)
        unpaidTabTV = findViewById(R.id.unpaidTabTV)
        pendingTabTV = findViewById(R.id.pendingTabTV)

        setupStatusTabs()

        adapter = RecentTransactionAdapter(
            transactionList,
            null,
            RecentTransactionAdapter.OnTransactionClickListener { transaction ->
                onTransactionTap(transaction)
            })
        recyclerView?.layoutManager = LinearLayoutManager(this)
        recyclerView?.adapter = adapter
    }

    private fun setupStatusTabs() {
        allTabTV?.let { setStatusTabSelected(it) }

        allTabTV?.setOnClickListener {
            selectedStatusTab = "All"
            allTabTV?.let { setStatusTabSelected(it) }
            refreshTransactions()
        }

        paidTabTV?.setOnClickListener {
            selectedStatusTab = "Paid"
            paidTabTV?.let { setStatusTabSelected(it) }
            refreshTransactions()
        }

        unpaidTabTV?.setOnClickListener {
            selectedStatusTab = "Unpaid"
            unpaidTabTV?.let { setStatusTabSelected(it) }
            refreshTransactions()
        }

        pendingTabTV?.setOnClickListener {
            selectedStatusTab = "Pending"
            pendingTabTV?.let { setStatusTabSelected(it) }
            refreshTransactions()
        }
    }

    private fun setStatusTabSelected(selectedTab: TextView) {
        allTabTV?.setBackgroundResource(0)
        paidTabTV?.setBackgroundResource(0)
        unpaidTabTV?.setBackgroundResource(0)
        pendingTabTV?.setBackgroundResource(0)
        selectedTab.setBackgroundResource(R.drawable.spinner_border_grey)
    }

    private fun refreshTransactions() {
        fetchTransactionsInRange(startDate, endDate)
    }

    private fun onTransactionTap(transaction: RecentTransaction?) {
        // Handle tap if needed
    }

    private fun getCurrentNickname() {
        val user = mAuth?.currentUserOrNull() ?: return
        val authId = user.id
        
        lifecycleScope.launch {
            try {
                val userDetails = DeclareDatabase.usersTable.select {
                    filter { eq("auth_id", authId) }
                }.decodeSingleOrNull<User>()
                
                currentNickname = userDetails?.username ?: ""
                currentUserNumericId = userDetails?.id
                
                // Refresh transactions once we have the numeric ID
                refreshTransactions()
            } catch (e: Exception) {
                Log.e("AllTransactions", "Error getting nickname: " + e.message)
            }
        }
    }

    private fun setupDateRangeSpinner() {
        val options = mutableListOf<String?>("This Month", "Last Month", "All", "Custom Date")
        val spinnerAdapter = SpinnerItemMonths(this, options)
        dateRangeSpinner?.adapter = spinnerAdapter
        setThisMonth()
        refreshTransactions()

        dateRangeSpinner?.onItemSelectedListener = object : android.widget.AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: android.widget.AdapterView<*>?, view: View?, position: Int, id: Long) {
                when (position) {
                    0 -> { setThisMonth(); refreshTransactions() }
                    1 -> { setLastMonth(); refreshTransactions() }
                    2 -> { setAllTime(); refreshTransactions() }
                    3 -> showDateRangePickerDialog(spinnerAdapter, options)
                }
            }
            override fun onNothingSelected(parent: android.widget.AdapterView<*>?) {}
        }
    }

    private fun setThisMonth() {
        val cal = Calendar.getInstance()
        cal.set(Calendar.DAY_OF_MONTH, 1)
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        startDate = cal.timeInMillis
        cal.set(Calendar.DAY_OF_MONTH, cal.getActualMaximum(Calendar.DAY_OF_MONTH))
        cal.set(Calendar.HOUR_OF_DAY, 23); cal.set(Calendar.MINUTE, 59); cal.set(Calendar.SECOND, 59)
        endDate = cal.timeInMillis
    }

    private fun setLastMonth() {
        val cal = Calendar.getInstance()
        cal.add(Calendar.MONTH, -1)
        cal.set(Calendar.DAY_OF_MONTH, 1)
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        startDate = cal.timeInMillis
        cal.set(Calendar.DAY_OF_MONTH, cal.getActualMaximum(Calendar.DAY_OF_MONTH))
        cal.set(Calendar.HOUR_OF_DAY, 23); cal.set(Calendar.MINUTE, 59); cal.set(Calendar.SECOND, 59)
        endDate = cal.timeInMillis
    }

    private fun setAllTime() {
        startDate = 0L
        endDate = Long.MAX_VALUE
    }

    private fun showDateRangePickerDialog(spinnerAdapter: SpinnerItemMonths, options: MutableList<String?>) {
        val safeStart = if (startDate == 0L || startDate == Long.MAX_VALUE)
            Calendar.getInstance().also { it.set(Calendar.DAY_OF_MONTH, 1) }.timeInMillis
        else startDate
        val safeEnd = if (endDate == Long.MAX_VALUE) Calendar.getInstance().timeInMillis else endDate

        val picker = MaterialDatePicker.Builder.dateRangePicker()
            .setTitleText("Select Date Range")
            .setSelection(Pair(safeStart, safeEnd))
            .build()
        picker.show(supportFragmentManager, "DATE_RANGE_PICKER")
        picker.addOnPositiveButtonClickListener { selection ->
            startDate = selection.first ?: safeStart
            endDate = (selection.second ?: selection.first ?: safeEnd) + 86400000 - 1
            val sdf = SimpleDateFormat("MMM dd", Locale.getDefault())
            options[3] = "${sdf.format(startDate)} - ${sdf.format(endDate)}"
            spinnerAdapter.notifyDataSetChanged()
            dateRangeSpinner?.setSelection(3)
            refreshTransactions()
        }
        picker.addOnCancelListener { dateRangeSpinner?.setSelection(0) }
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun fetchTransactionsInRange(start: Long, end: Long) {
        if (currentUserNumericId == null && mAuth?.currentUserOrNull() != null) {
            // Wait for user details to load
            return
        }

        loadingProgressBar?.visibility = View.VISIBLE
        emptyStateLayout?.visibility = View.GONE
        transactionList.clear()
        adapter?.notifyDataSetChanged()

        lifecycleScope.launch {
            try {
                val transactions = DeclareDatabase.transactionsTable.select().decodeList<Transaction>()

                for (transaction in transactions) {
                    val timestamp = parseCreatedAt(transaction.createdAt)
                    
                    if (timestamp in start..end && isUserInvolved(transaction, currentNickname)) {
                        if (!matchesStatusFilter(transaction, selectedStatusTab)) {
                            continue
                        }

                        val cal = Calendar.getInstance()
                        cal.timeInMillis = timestamp
                        val monthName = cal.getDisplayName(Calendar.MONTH, Calendar.LONG, Locale.getDefault())
                        val year = cal.get(Calendar.YEAR).toString()
                        val day = cal.get(Calendar.DAY_OF_MONTH).toString()
                        val monthYear = "$monthName-$year"

                        val displayDate = "$monthName - $day"
                        val fullDateWithYear = "$monthName $day, $year"
                        val timeKey = SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(cal.time)
                        val sortDateTime = "$year-$monthName-$day $timeKey"

                        val transactionType = transaction.transactionType
                        val details = transaction.transactionDetail
                        val paymentAmount = transaction.paymentAmount
                        val paymentAmountStr = CurrencyUtils.formatAmountWithCurrency(paymentAmount)
                        val iconResource = getIconForTransactionType(transactionType)

                        val payorsList = (transaction.payorsDisplayNames ?: transaction.contributors)?.map { it as String? }?.toMutableList()
                        val payorUserIds = transaction.contributors?.map { it as String? }?.toMutableList()
                        val amountsPaidList = transaction.amountPaidList?.map { it as Double? }?.toMutableList()
                        val totalIndividualPayment = transaction.individualPayment

                        var createdBy = transaction.posterDisplayName
                        if (createdBy.isNullOrEmpty()) {
                            createdBy = transaction.usernamePost
                        }
                        val createdByUserId = transaction.usernamePost

                        val recentTrans = RecentTransaction(
                            transaction.id,
                            displayDate, transactionType, details, paymentAmountStr,
                            iconResource, sortDateTime, timestamp, payorsList, payorUserIds,
                            amountsPaidList, totalIndividualPayment, fullDateWithYear,
                            createdBy, createdByUserId, monthYear, day, timeKey
                        )
                        transactionList.add(recentTrans)
                    }
                }

                transactionList.sortWith { t1, t2 ->
                    t2.timestamp.compareTo(t1.timestamp)
                }

                adapter?.notifyDataSetChanged()
                adapter?.preloadAllImages(this@AllTransactionsActivity)

                val count = transactionList.size
                transactionCountTextView?.text = "$count ${if (count == 1) "transaction" else "transactions"}"

                if (transactionList.isEmpty()) {
                    emptyStateLayout?.visibility = View.VISIBLE
                    recyclerView?.visibility = View.GONE
                } else {
                    emptyStateLayout?.visibility = View.GONE
                    recyclerView?.visibility = View.VISIBLE
                }
            } catch (e: Exception) {
                Log.e("AllTransactions", "Error loading transactions: " + e.message)
            } finally {
                loadingProgressBar?.visibility = View.GONE
            }
        }
    }

    private fun parseCreatedAt(createdAt: String?): Long {
        if (createdAt == null) return 0L
        val formats = arrayOf(
            "yyyy-MM-dd'T'HH:mm:ss.SSSXXX",
            "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
            "yyyy-MM-dd'T'HH:mm:ss",
            "yyyy-MM-dd HH:mm:ss"
        )
        for (format in formats) {
            try {
                val sdf = SimpleDateFormat(format, Locale.getDefault())
                val date = sdf.parse(createdAt)
                if (date != null) return date.time
            } catch (e: Exception) {
                // Try next format
            }
        }
        return 0L
    }

    private fun matchesStatusFilter(
        transaction: Transaction,
        statusFilter: String?
    ): Boolean {
        if ("All".equals(statusFilter, ignoreCase = true)) return true

        val paidAmounts = transaction.amountPaidList
        val totalToPay = transaction.individualPayment

        if (paidAmounts.isNullOrEmpty()) {
            return "Unpaid".equals(statusFilter, ignoreCase = true)
        }

        var allPaid = true
        var allUnpaid = true

        for (paid in paidAmounts) {
            if (paid < totalToPay) {
                allPaid = false
            }
            if (paid > 0) {
                allUnpaid = false
            }
        }

        val status = if (allPaid) "Paid" else if (allUnpaid) "Unpaid" else "Pending"
        return status.equals(statusFilter, ignoreCase = true)
    }

    private fun getIconForTransactionType(transactionType: String?): Int {
        return when (transactionType) {
            "Electricity" -> R.drawable.lightning_bolt
            "Water" -> R.drawable.faucet
            "Rent" -> R.drawable.house
            "Internet" -> R.drawable.internet
            "Online Shopping" -> R.drawable.online_shopping
            "Travel" -> R.drawable.travel
            "Groceries" -> R.drawable.groceries
            "Foods" -> R.drawable.hamburger
            "House Necessity" -> R.drawable.necessities
            "Transportation" -> R.drawable.vehicles
            else -> R.drawable.others
        }
    }

    private fun isUserInvolved(
        transaction: Transaction?,
        usernameOrUserId: String?
    ): Boolean {
        if (transaction == null) return false
        
        // 1. Check by numeric ID (Best for consistency with storage/other models)
        if (currentUserNumericId != null) {
            if (transaction.creatorId == currentUserNumericId) return true
            if (transaction.contributors?.contains(currentUserNumericId.toString()) == true) return true
        }
        
        // 2. Check by Auth UID (Fallback for legacy data)
        val currentAuthId = mAuth?.currentUserOrNull()?.id
        if (currentAuthId != null) {
            if (transaction.usernamePost == currentAuthId) return true
            if (transaction.contributors?.contains(currentAuthId) == true) return true
        }
        
        // 3. Check by Nickname (Fallback for names stored in contributors)
        if (usernameOrUserId != null) {
            if (usernameOrUserId == transaction.posterDisplayName) return true
            if (transaction.payorsDisplayNames?.contains(usernameOrUserId) == true) return true
            if (transaction.contributors?.contains(usernameOrUserId) == true) return true
        }

        return false
    }
}
